### PR TITLE
Added font glyph api to dmSDK

### LINKED ISSUE #9376: Add support for runtime font glyph generation
We want a way to ship a game with few (or no) glyphs.
Instead, we'd like it to be able to load glyphs from a .ttf file, and generate glyphs at runtime.

Extension:
https://github.com/defold/extension-fontgen

### PR DESCRIPTION
**Pull request by JCash**
We've added a api to our dmSDK for manipulating font resources.
This allows for writing extensions such as [extension-font](https://github.com/defold/extension-font), making it possible to avoid storing large amounts of pre rendered glyph images in the default game, and instead generate them at runtime.

The api has been added to `dmsdk/gamesys/resources/res_font.h`.

```c++
#include <dmsdk/gamesys/resources/res_font.h>

dmRender::HFont    dmGameSystem::ResFontGetHandle(FontResource* font);
dmResource::Result dmGameSystem::ResFontGetInfo(FontResource* font, FontInfo* info);
dmResource::Result dmGameSystem::ResFontGetCacheCellInfo(FontResource* font, uint32_t* width, uint32_t* height, uint32_t* max_ascent);
bool               dmGameSystem::ResFontHasGlyph(FontResource* font, uint32_t codepoint);
dmResource::Result dmGameSystem::ResFontAddGlyph(FontResource* font, uint32_t codepoint, FontGlyph* glyph, void* imagedata, uint32_t imagedatasize);
dmResource::Result dmGameSystem::ResFontRemoveGlyph(FontResource* font, uint32_t codepoint);
```

Fixes #9376 

## PR checklist

* [ ] Code
	* [ ] Add engine and/or editor unit tests.
	* [ ] New and changed code follows the overall code style of existing code
	* [ ] Add comments where needed
* [ ] Documentation
	* [ ] Make sure that API documentation is updated in code comments
	* [ ] Make sure that manuals are updated (in github.com/defold/doc)
* [ ] Prepare pull request and affected issue for automatic release notes generator
	* [ ] Pull request - Write a message that explains what this pull request does. What was the problem? How was it solved? What are the changes to APIs or the new APIs introduced? This message will be used in the generated release notes. Make sure it is well written and understandable for a user of Defold.
	* [ ] Pull request - Write a pull request title that in a sentence summarises what the pull request does. Do not include "Issue-1234 ..." in the title. This text will be used in the generated release notes.
	* [ ] Pull request - Link the pull request to the issue(s) it is closing. Use on of the [approved closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).
	* [ ] Affected issue - Assign the issue to a project. Do not assign the pull request to a project if there is an issue which the pull request closes.
	* [ ] Affected issue - Assign the "breaking change" label to the issue if introducing a breaking change.
	* [ ] Affected issue - Assign the "skip release notes" is the issue should not be included in the generated release notes.

----------

Example of a well written PR description:

1. Start with the user facing changes. This will end up in the release notes.
1. Add one of the GitHub approved closing keywords
1. Optionally also add the technical changes made. This is information that might help the reviewer. It will not show up in the release notes. Technical changes are identified by a line starting with one of these:
   1. `### Technical changes` 
   1. `Technical changes:`
   2. `Technical notes:`

```
There was a anomaly in the carbon chroniton propeller, introduced in version 8.10.2. This fix will make sure to reset the phaser collector on application startup.

Fixes #1234

### Technical changes
* Pay special attention to line 23 of phaser_collector.clj as it contains some interesting optimizations
* The propeller code was not taking into account a negative phase.
```

### CODE REVIEW

File: engine/render/proto/render/font_ddf.proto [hunk 1]
```diff
@@ -42,20 +45,23 @@ message FontDesc
...
-        optional int32  x                 = 7 [default = 0];
-        optional int32  y                 = 8 [default = 0];
-        optional uint64 glyph_data_offset = 9;
-        optional uint64 glyph_data_size   = 10;
+        optional uint32 ascent            = 5 [default = 0]; // WHY 32 bits???
+        optional uint32 descent           = 6 [default = 0]; // WHY 32 bits???
+        optional int32  x                 = 7 [default = 0]; // only used in editor preview
+        optional int32  y                 = 8 [default = 0]; // only used in editor preview
+        optional uint64 glyph_data_offset = 9; // WHY 64 bits???
+        optional uint64 glyph_data_size   = 10; // 32 bits should suffice?
```
Comment (JCash) [inline engine/render/proto/render/font_ddf.proto:64]: Future memory optimizations.

File: engine/render/proto/render/font_ddf.proto [hunk 2]
```diff
@@ -65,32 +71,45 @@ message GlyphBank
 
     optional float             max_ascent            = 5;
     optional float             max_descent           = 6;
-    optional FontTextureFormat image_format          = 7  [default = TYPE_BITMAP];
+    optional float             max_advance           = 7;
+    optional float             max_width             = 8;
+    optional float             max_height            = 9;
```
Comment (JCash) [inline engine/render/proto/render/font_ddf.proto:76]: More dimensions of the font.

File: engine/render/src/dmsdk/render/render.h [hunk 3]
```diff
@@ -55,10 +55,14 @@ namespace dmRender
     /*#
      * Font map handle
      * @typedef
-     * @name HFontMap
+     * @name HFont
      */
+    typedef struct FontMap* HFont;
+
+    // Old typedef, used internally. We want to migrate towards HFont
     typedef struct FontMap* HFontMap;
```
Comment (JCash) [inline engine/render/src/dmsdk/render/render.h:63]: A quick "rename" (the internal HFontMap is used _a lot_, so I chose to wait with that)

File: engine/render/src/render/font_renderer.cpp [hunk 4]
```diff
@@ -57,6 +58,18 @@ namespace dmRender
...
+    {
+        // NOTE: The struct *must* be 16-bytes aligned due to SIMD operations.
+        float m_Position[4];
+        float m_UV[2];
+        float m_FaceColor[4];
+        float m_OutlineColor[4];
+        float m_ShadowColor[4];
+        float m_SdfParams[4];
+        float m_LayerMasks[3];
+    };
```
Comment (JCash) [inline engine/render/src/render/font_renderer.cpp:71]: Moved from header.

File: engine/render/src/render/font_renderer.cpp [hunk 5]
```diff
@@ -202,18 +235,64 @@ namespace dmRender
...
+
+            CacheGlyph* glyph = &font_map->m_Cache[i];
+            glyph->m_Glyph = 0;
+            glyph->m_Frame = 0;
+
+            // We calculate these only once
+            uint32_t col = i % font_map->m_CacheColumns;
+            uint32_t row = i / font_map->m_CacheColumns;
+            glyph->m_X = col * font_map->m_CacheCellWidth;
+            glyph->m_Y = row * font_map->m_CacheCellHeight;
```
Comment (JCash) [inline engine/render/src/render/font_renderer.cpp:276]: Each cell in the cache has a predetermined location in the texture.

File: engine/resource/src/dmsdk/resource/resource.h [hunk 6]
```diff
@@ -210,10 +210,30 @@ ResourceResult ResourceGetByHash(HResourceFactory factory, dmhash_t name, void**
...
+
+/*#
+ * Get resource descriptor from resource (name)
+ * @name GetDescriptorByHash
+ * @param factory [type: HResourceFactory] Factory handle
+ * @param path_hash [type: dmhash_t] Resource path hash
+ * @param descriptor [type: HResourceDescriptor*] Returned resource descriptor
+ * @return result [type: ResourceResult]  RESULT_OK on success
+ */
+ResourceResult ResourceGetDescriptorByHash(HResourceFactory factory, dmhash_t path_hash, HResourceDescriptor* descriptor);
```
Comment (JCash) [inline engine/resource/src/dmsdk/resource/resource.h:231]: Some missing functions, to be able to check the type in the extension.

File: engine/resource/src/dmsdk/resource/resource_gen.hpp [hunk 7]
```diff
@@ -170,13 +170,35 @@ namespace dmResource
...
+
+    /*# 
+     * Release resource
+     * @name Release
+     * @language C++
+     * @param factory [type:HResourceFactory] Factory handle
+     * @param resource [type:void*] Resource pointer
+     * @note Decreases ref count by 1. If it reaches 0, the resource destroy function is called.
+     */
+    void Release(HFactory factory, void * resource);
```
Comment (JCash) [inline engine/resource/src/dmsdk/resource/resource_gen.hpp:190]: Missing from the header (buf from last update)

File: engine/resource/src/resource.cpp [hunk 8]
```diff
@@ -1540,3 +1542,69 @@ const char* ResultToString(Result r)
 }
 
 } // namespace dmResource
+
+// The C interface
+void ResourceRegisterReloadedCallback(HResourceFactory factory, FResourceReloadedCallback callback, void* user_data)
```
Comment (JCash) [inline engine/resource/src/resource.cpp:1547]: Unresovled symbols. (bug from last update)

File: engine/resource/wscript [hunk 9]
```diff
@@ -83,6 +84,9 @@ def _wait_for_server(self, local_ip, port):
             print('Waiting for HTTP testserver to start...')
             sys.stdout.flush()
             time.sleep(0.5)
+        except Exception as e:
+            self.fatal('Failed to start test server at http://%s:%d: %s: %s, Is there another test server running?' % (local_ip, port, type(e), e))
```
Comment (JCash) [inline engine/resource/wscript:88]: Nicer output for newcomers (Python/waf is too callstack-prone)

File: engine/gamesys/src/gamesys/resources/res_font.cpp [hunk 10]
```diff
@@ -0,0 +1,424 @@
...
+        DynamicGlyph** dynglyphp = resource->m_DynamicGlyphs.Get(codepoint);
+        if (dynglyphp)
+        {
+
+            DynamicGlyph* dynglyph = *dynglyphp;
+            *out_size = dynglyph->m_DataSize;
+            *out_width = dynglyph->m_DataImageWidth;
+            *out_height = dynglyph->m_DataImageHeight;
+            *out_compression = dynglyph->m_Data[0];
+            return dynglyph->m_Data+1;
```
[the remark this answers is not in the document]
Comment (Jhonnyg) [reply engine/gamesys/src/gamesys/resources/res_font.cpp:136]: The +1 thing here might need an explanation in code, I don't really understand what its for? I'm guessing the DynamicGlyph* dynglyph here points to a header for the glyph data, and dynglyph +1 is the payload?

File: engine/gamesys/src/gamesys/resources/res_font.cpp [hunk 11]
```diff
@@ -0,0 +1,426 @@
...
+        g.m_Width      = inglyph->m_Width;
+        g.m_Advance    = inglyph->m_Advance;
+        g.m_LeftBearing= inglyph->m_LeftBearing;
+        g.m_Ascent     = inglyph->m_Ascent;
+        g.m_Descent    = fabs(inglyph->m_Descent);
+        // Redundant in this setup
+        // g.m_X;
+        // g.m_Y;
+        // g.m_GlyphDataOffset;
+        // g.m_GlyphDataSize;
```
Comment (Jhonnyg) [inline engine/gamesys/src/gamesys/resources/res_font.cpp:355]: Are these not used anymore, or why are they redundant?

**Review by Jhonnyg — DISMISSED** [review]

File: engine/gamesys/src/gamesys/resources/res_font.cpp (hunk 11 above)
Comment (JCash) [reply engine/gamesys/src/gamesys/resources/res_font.cpp:355]: the X/Y was never used at runtime. only in the editor.

And, the dynamic system doesn't produce the data in the same way, so the glyph data is different and handled explicitly.

**Pull request merged**